### ISSUE TITLE
BetterBibTeX citation key when accessing group library via server API (pyzotero)

### ISSUE THREAD

**Issue by fabbra** [retorquere/zotero-better-bibtex#1497]
For the purpose of automated bibliography generation of my research group I am accessing a Zotero group library via the server API (using pyzotero). To generate the bibliography I will be using LaTeX and the automatically generated (and updated) *.bib file from Zotero BetterBibTex.

My issue lies in **finding the correct association/mapping between 1) the BetterBibTex citation key and 2) the library key Zotero** for each item in my library. This was [initially described in the Zotero forum](https://forums.zotero.org/discussion/82437/betterbibtex-citation-key-when-accessing-group-library-via-server-api-pyzotero) but I am not yet satisfied with the possible solutions:

a) My current workaround consists in exporting a BIB file (via BetterBibTex) to do the actual bibliography AND a **JSON file** (via BetterBibTex JSON). The latter is **parsed to determine** - for every library item - the 1) BetterBibTex **citation key** and 2) its Zotero **library key**. 
Until recently 2) was obtained directly from the variable "_key_". However, this one seems to have disappeared (with the most recent version of BetterBibTex or configuration changes that I've made without being aware of) such that the current workaround consists in reading the "_uri_" variable and extracting the key as the string after the last "/".

b) As suggested [in the Zotero forum](https://forums.zotero.org/discussion/82437/betterbibtex-citation-key-when-accessing-group-library-via-server-api-pyzotero) one could **pin the citation keys** such that these can be **read directly from the "_extra_" variable** of each library item using the **server API** (via pyzotero).
However, my library will be accessed by various people who can add new library items and I unfortunately cannot guarantee that all of them have enabled [the hidden `autoPin` preference](https://retorque.re/zotero-better-bibtex/installation/preferences/hidden-preferences/#autopin), which should automatically pin citekeys. Besides, I want this to be a fully automatic process, avoiding any manual interaction. 
Or did I misunderstand the principle of the `autoPin` preference and would it work for all items of my group library even if some users (adding new library items) do not have it enabled?

Is there a better way to achieve this association/mapping between 1) and 2) ?

Thank you for your help and for your efforts in developing this amazing tool!
Fab

PS: I'm running the latest version of BetterBibTex (5.2.21) and Zotero (5.0.85) on Win64.

**Comment by retorquere** [comment]
Would you happen to have both the libraryID and the key, or just the key? I think Zotero assures that keys are unique within a group/library but not necessarily across all groups/libraries. If you have the item URI, that would also suffice, because that contains the libraryID and key.

**Comment by fabbra** [comment]
Not sure I fully understand your request. 
The current library I am referring to is a private one. If you want access to test I can create a public library for testing. Is this what you want?

**Comment by retorquere** [comment]
No, I don't need access, let me restate. You say you want to look up the BBT citekey based on the zotero library key. I need to make sure w're using the teminology in the same way, so:

as far as I know, there's no concept of "library key" to zotero. There are library IDs, which are numeric, and item (and collection) keys, which are alphanumeric. I am guessing that by "library key" you mean "item key" because you can't sensibly look up a citekey for a whole library.

So if your question is, in zotero terminology, "can I look up the BBT citekey based on the zotero item key", my answer is "yes, but probably only reliably if you also know the library ID". If you have only a private library, that still has a library ID; if you were to right-click on a collection in zotero and select "download better bibtex export", the first URL will show the library ID and the collection key. In your case, the library ID is probably `1`.

How do you get the item key?

**Comment by fabbra** [comment]
Okay, sorry for having caused misunderstandings by not using the correct terminology.

Yes, that's it: I would like to lookup 1) the "BBT citekey" based on 2) the "item key" (which I mistakenly called "library key").

I am able to show the URL mentioned by you using "download better bibtex export" for a collection. However, when trying to access the URL in my browser I get a `Request not allowed` message.

So how will this help me to get the link between "BBT citekey" and "item key"?

**Comment by retorquere** [comment]
No apologies required or expected, I just needed to set the vocabulary so we can communicate more effectively.

You can't access that URL in the browser; Zotero forbids it. But just knowing the URL will give you the library ID, you don't need its contents.

What you want can be done, but not by calling the public Zotero API. I could give you a local API that calls into a running Zotero which can give you the citekey based on the itemKey.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.2.21.6403](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6403.xpi) ("fixes #1497")

Install in Zotero by downloading [test build 5.2.21.6403](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6403.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by retorquere** [comment]
On 6403, you can call the local API using

```
curl http://localhost:23119/better-bibtex/json-rpc -X POST -H "Content-Type: application/json" -H "Accept: application/json" --data-binary '{"jsonrpc": "2.0", "method": "items.citationkey", "params": [["6STQAJ5F", "1:KG9BBSIH", "Nope"]] }'
```

which on my test library yields

```
{"jsonrpc":"2.0","result":{"6STQAJ5F":"ShelahStabletheories1969","1:KG9BBSIH":"ShelahNoteminmaxproblem1969","Nope":null},"id":null}
```
You give it the list of item keys to look up, and you get a JSON dict back that maps the keys you gave to the corresponding citation key. The keys you provide must be in the form `<libraryID>:<itemKey>` or `<itemKey>`; in the latter case, it will search for the keys in your personal library. If an item cannot be found, the value `null` will be returned instead of the citation key.

**Comment by fabbra** [comment]
Works like a charm!

One more question: If I add more group libraries, do I risk that the `<libraryID>` of the "first" library (i.e. my _library of interest_) changes or will it always remain the same independent of other group libraries?

**Comment by retorquere** [comment]
It might change, but not to an ID that's in use anywhere else. If you just omit it, it will reliably pick out your personal library.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.2.21.6404](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6404.xpi) ("fixes #1497")

Install in Zotero by downloading [test build 5.2.21.6404](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6404.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.2.21.6407](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6407.xpi) ("fixes #1497")

Install in Zotero by downloading [test build 5.2.21.6407](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6407.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.2.21.6409](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6409.xpi) ("fixes #1497")

Install in Zotero by downloading [test build 5.2.21.6409](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6409.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.2.21.6418](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6418.xpi) ("fixes #1497")

Install in Zotero by downloading [test build 5.2.21.6418](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6418.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.2.21.6424](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6424.xpi) ("fixes #1497")

Install in Zotero by downloading [test build 5.2.21.6424](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.21.6424.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".
